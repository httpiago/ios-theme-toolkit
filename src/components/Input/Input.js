import React from 'react'
import PropTypes from 'prop-types'
import { mapClassList } from '../../Utils'

/**
 * Input component.
 *
 * @since 0.1.0
 * @see https://httpiago.github.io/ios-theme-toolkit/#/components/Input
 */
export default function Input({ htmlType, autosize, fill, error, round, block, className: aditionalClasses, onEnter, onCancel, ...rest }) {

  const prefix = 'input-text'
  const classes = mapClassList({
    [`${prefix}--fill`]: fill,
    [`${prefix}--round`]: round,
    [`${prefix}--block`]: block,
    [`${prefix}--error`]: error
  })

  const handleKeyDown = (event) => {
    if (event.keyCode === 13 && onEnter) onEnter(event)
    else if (event.keyCode === 27 && onCancel) onCancel(event)
  }

  if (htmlType === 'textarea') {
    return (
      <textarea
        className={`${prefix} ${classes} ${aditionalClasses}`}
        autosize={String(autosize)}
        onKeyDown={handleKeyDown}
        {...rest}
      />
    )
  } else {
    return (
      <input
        type={htmlType}
        className={`${prefix} ${classes} ${aditionalClasses}`}
        onKeyDown={handleKeyDown}
        {...rest}
      />
    )
  }
}

Input.defaultProps = {
  htmlType: 'text',
  round: false,
  block: false,
  autosize: false,
  className: '',

  onEnter: function(event) {},
  onCancel: function(event) {},
  onFocus: function(event) {},
  onBlur: function(event) {},
}

Input.propTypes = {
  /** Max length of input */
  maxLength: PropTypes.oneOfType([ PropTypes.string, PropTypes.number ]),
  /** Disable or not the input. */
  disabled: PropTypes.bool,
  /** The input value for CONTROLLED components. */
  value: PropTypes.string,
  /** The initial input value for UNCONTROLLED components. */
  defaultValue: PropTypes.string,
  /** Default value that is shown when input is empty. */
  placeholder: PropTypes.string,
  /** Height autosize feature. (not finished) */
  autosize: PropTypes.oneOfType([ PropTypes.bool, PropTypes.string ]),
  /** Fit input width to its parent width. */
  block: PropTypes.bool,
  /** Round edges. */
  round: PropTypes.bool,
  /** Aditional classes. */
  className: PropTypes.string,
  /** Set the original html type of input tag */
  htmlType: PropTypes.oneOf([ 'text', 'textarea', 'password', 'number', 'email', 'url', 'search', 'tel', 'time', 'date', 'file' ]),
  /** Function that will called when user press Enter key */
  onEnter: PropTypes.func,
  /** Function that will called when user press ESC key */
  onCancel: PropTypes.func,

  onFocus: PropTypes.func,
  onChange: PropTypes.func,
  onBlur: PropTypes.func,
  onKeyDown: PropTypes.func,
  onKeyUp: PropTypes.func
}
