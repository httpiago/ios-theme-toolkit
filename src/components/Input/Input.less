@import '../../customizations.less';

// Default styles
.input-text {
  -webkit-box-sizing: border-box;
  box-sizing: border-box;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: transparent;
  border: none;
  border-radius: @base-border-radius;
  outline: 0;
  display: block;
  padding: 0 0;
  margin: 0;
  line-height: 36px;
  height: 36px;
  resize: none;
  font-family: @font-family;
  font-size: @font-size + 2;
  color: #333;
}

// Input com textare
textarea.input-text {
  resize: vertical;
}

// Caixa de texto com fundo
.input-text--fill {
  background-color: rgba(0, 0, 0, .12);
  padding: 0 12px;

  // Somente dar foco quando o input tiver a classe input--fill
  &:focus {
    background-color: rgba(0, 0, 0, .2);
    box-shadow: 0 0 0 3px fade(@gray-color, 20%);
  }
}

// Caixa de texto que ocupa todo o espa√ßo horizontal
.input-text--block {
  display: block;
  width: 100%;
}

// Caixa de texto com cantos arredondados
.input-text--round {
  border-radius: 50px;
}

// Caixa de texto desabilitada
.input-text[disabled] {
  opacity: .56;
  cursor: default;
}

// Caixa de texto com propriedade readOnly
.input-text[readonly] {
  cursor: no-drop;
}

// Input com error
.input-text--error {
  background-color: fade(@red-color, 12%);
  color: @red-color;

  &.input-text--fill:focus {
    box-shadow: 0 0 0 3px fade(@red-color, 50%);
  }

  & + .error-msg {
    font-size: 14px;
    color: @red-color;
  }
}
