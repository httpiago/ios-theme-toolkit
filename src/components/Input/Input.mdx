---
name: Input
route: /components/Input
menu: Componentes
---

import { Playground, PropsTable } from 'docz'
import { Input } from '../../index.js'
import '../../../dist/styles.css'

# Input

To use this component, simply import it into any file and pass the [modifiers](#/#props) in the properties.

**NOTE: This package does not provide any form of data validation, I recommend you take a look at [formik](https://github.com/jaredpalmer/formik).**

```jsx
import { Input } from 'ios-theme-toolkit'

export default () => (
  <Input value="Hello world" />
)
```

## Examples

<Playground>
  <span className="desc">Without filling:</span>
  <Input
    placeholder="Type your name..."
    onChange={(e) => console.log('input changed', e.target.value)}
  />

  <span className="desc">With fill:</span>
  <Input
    fill
    placeholder="Type your email"
    onEnter={e => alert('your email: ' + e.target.value)}
    onCancel={e => alert('canceled')}
  />

  <span className="desc">Textarea:</span>
  <Input
    htmlType="textarea"
    fill
    block
    placeholder="Type your text here..."
    defaultValue="Lorem ipsum dolor sit amet, consectetur adipiscing elit. Fusce in vulputate enim. Duis aliquet imperdiet viverra. Suspendisse vestibulum urna eget nisl sollicitudin pulvinar. Suspendisse porta nibh diam, varius accumsan massa sodales sed."
    style={{ height: 140 }}
  />

  <span className="desc">Rounded edges:</span>
  <Input
    fill
    round
    placeholder="Input with rounded edges"
  />

  <span className="desc">Invalid input:</span>
  <Input
    fill
    error
    required
    defaultValue="Input invalid"
  />
  <span className="error">Digite um valor v√°lido</span>
  <span className="desc">Disabled:</span>
  <Input
    fill
    disabled
    placeholder="Type your new password"
  />

  <span className="desc">Read-only:</span>
  <Input
    fill
    disabled
    readOnly
    defaultValue="Read only input"
  />

  <style>{".react-live-preview .desc { display: block; font-weight: bold; margin-top: 20px }"}</style>
</Playground>

## Props

<PropsTable of={ Input } />

All the properties which are not listed above will be transferred to the \<input\> (or \<Textarea\>) tag, so, It accepts all props which native input support.
