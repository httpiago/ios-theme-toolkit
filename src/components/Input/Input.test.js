/**
 * TEST FILE
 * Please try writing the tests for all component behaviors,
 * manually call the functions within the class, change the properties...
 *
 * @see https://jestjs.io/docs/en/expect
 * @see https://github.com/airbnb/enzyme/tree/master/docs/api
 */
import React from 'react'
import { shallow } from 'enzyme'
import Input from './Input'

const COMPONENT = shallow(
  <Input defaultValue="test value" className="aditionalClass" />
)

describe('Test Input component', () => {
  it('should render correctly based on last snapshot', () => {
    expect(COMPONENT).toMatchSnapshot()
  })

  it('should update correctly', () => {
    COMPONENT.setProps({ block: true })

    expect(COMPONENT.hasClass('input-text--block')).toEqual(true)
  })

  it('should has aditional classes', () => {
    expect(COMPONENT.hasClass('aditionalClass')).toEqual(true)
  })

  it('should return a textarea tag when prop "htmlType" are defined as "textarea"', () => {
    const TEXTAREACOMP = shallow(
      <Input htmlType="textarea" defaultValue="" />
    )

    expect(TEXTAREACOMP.type()).toEqual('textarea')
  })

  var enterPressed = false
  var escPressed = false
  var fakeStateInputValue = 'initial content'

  const INTERACTIVEINPUT = shallow(
    <Input
      value={fakeStateInputValue}
      onChange={(e) => { fakeStateInputValue = e.target.value }}
      onEnter={() => { enterPressed = true }}
      onCancel={() => { escPressed = true }}
    />
  )

  it('should be a controlled component', () => {
    // Simulate user input
    INTERACTIVEINPUT.setProps({ value: 'new value test' })

    expect(INTERACTIVEINPUT.html()).toContain('value="new value test"')
  })

  it('should callback when user press enter key', () => {
    // Simulate that user pressed enter
    INTERACTIVEINPUT.simulate('keyDown', { keyCode: 13 })

    expect(enterPressed).toBeTruthy()

  })

  it('should callback when user press esc key', () => {
    // Simulate that user pressed esc
    INTERACTIVEINPUT.simulate('keyDown', { keyCode: 27 })

    expect(escPressed).toBeTruthy()
  })
})
