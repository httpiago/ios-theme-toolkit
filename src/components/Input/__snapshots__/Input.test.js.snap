// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Input component should render correctly based on last snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Input
    autosize={false}
    block={false}
    className="aditionalClass"
    defaultValue="test value"
    htmlType="text"
    onBlur={[Function]}
    onCancel={[Function]}
    onEnter={[Function]}
    onFocus={[Function]}
    round={false}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "className": "input-text  aditionalClass",
      "defaultValue": "test value",
      "onBlur": [Function],
      "onFocus": [Function],
      "onKeyDown": [Function],
      "type": "text",
    },
    "ref": null,
    "rendered": null,
    "type": "input",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "className": "input-text  aditionalClass",
        "defaultValue": "test value",
        "onBlur": [Function],
        "onFocus": [Function],
        "onKeyDown": [Function],
        "type": "text",
      },
      "ref": null,
      "rendered": null,
      "type": "input",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
